File: java/client/src/test/java/org/apache/rocketmq/client/java/impl/consumer/PushSubscriptionSettingsTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.rocketmq.client.java.impl.consumer;

import apache.rocketmq.v2.ClientType;
import apache.rocketmq.v2.CustomizedBackoff;
import apache.rocketmq.v2.FilterType;
import apache.rocketmq.v2.Settings;
import apache.rocketmq.v2.Subscription;
import apache.rocketmq.v2.SubscriptionEntry;
import com.google.protobuf.util.Durations;
import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.apache.rocketmq.client.apis.consumer.FilterExpression;
import org.apache.rocketmq.client.apis.consumer.FilterExpressionType;
import org.apache.rocketmq.client.java.message.protocol.Resource;
import org.apache.rocketmq.client.java.misc.ClientId;
import org.apache.rocketmq.client.java.tool.TestBase;
import org.junit.Assert;
import org.junit.Test;

public class PushSubscriptionSettingsTest extends TestBase {

    @Test
    public void testToProtobuf() {
        Resource groupResource = new Resource(FAKE_CONSUMER_GROUP_0);
        ClientId clientId = new ClientId();
        Map<String, FilterExpression> subscriptionExpression = new HashMap<>();
        subscriptionExpression.put(FAKE_TOPIC_0, new FilterExpression());
        final Duration requestTimeout = Duration.ofSeconds(3);
        final PushSubscriptionSettings pushSubscriptionSettings = new PushSubscriptionSettings(clientId,
            fakeEndpoints(), groupResource, requestTimeout, subscriptionExpression);
        final Settings settings = pushSubscriptionSettings.toProtobuf();
        Assert.assertEquals(settings.getClientType(), ClientType.PUSH_CONSUMER);
        Assert.assertEquals(settings.getRequestTimeout(), Durations.fromNanos(requestTimeout.toNanos()));
        Assert.assertTrue(settings.hasSubscription());
        final Subscription subscription = settings.getSubscription();
        Assert.assertEquals(subscription.getGroup(),
            apache.rocketmq.v2.Resource.newBuilder().setName(FAKE_CONSUMER_GROUP_0).build());
        Assert.assertFalse(subscription.getFifo());
        final List<SubscriptionEntry> subscriptionsList = subscription.getSubscriptionsList();
        Assert.assertEquals(subscriptionsList.size(), 1);
        final SubscriptionEntry subscriptionEntry = subscriptionsList.get(0);
        Assert.assertEquals(subscriptionEntry.getExpression().getType(), FilterType.TAG);
        Assert.assertEquals(subscriptionEntry.getTopic(),
            apache.rocketmq.v2.Resource.newBuilder().setName(FAKE_TOPIC_0).build());
    }

    @Test
    public void testToProtobufWithSqlExpression() {
        Resource groupResource = new Resource(FAKE_CONSUMER_GROUP_0);
        ClientId clientId = new ClientId();

        Map<String, FilterExpression> subscriptionExpression = new HashMap<>();
        subscriptionExpression.put(FAKE_TOPIC_0, new FilterExpression("(a > 10 AND a < 100) OR (b IS NOT NULL AND "
            + "b=TRUE)", FilterExpressionType.SQL92));
        final Duration requestTimeout = Duration.ofSeconds(3);
        final PushSubscriptionSettings pushSubscriptionSettings = new PushSubscriptionSettings(clientId,
            fakeEndpoints(), groupResource, requestTimeout, subscriptionExpression);
        final Settings settings = pushSubscriptionSettings.toProtobuf();
        Assert.assertEquals(settings.getClientType(), ClientType.PUSH_CONSUMER);
        Assert.assertEquals(settings.getRequestTimeout(), Durations.fromNanos(requestTimeout.toNanos()));
        Assert.assertTrue(settings.hasSubscription());
        final Subscription subscription = settings.getSubscription();
        Assert.assertEquals(subscription.getGroup(),
            apache.rocketmq.v2.Resource.newBuilder().setName(FAKE_CONSUMER_GROUP_0).build());
        Assert.assertFalse(subscription.getFifo());
        final List<SubscriptionEntry> subscriptionsList = subscription.getSubscriptionsList();
        Assert.assertEquals(subscriptionsList.size(), 1);
        final SubscriptionEntry subscriptionEntry = subscriptionsList.get(0);
        Assert.assertEquals(subscriptionEntry.getExpression().getType(), FilterType.SQL);
        Assert.assertEquals(subscriptionEntry.getTopic(),
            apache.rocketmq.v2.Resource.newBuilder().setName(FAKE_TOPIC_0).build());
    }

    @Test
    public void testSync() {
        com.google.protobuf.Duration duration0 = Durations.fromSeconds(1);
        com.google.protobuf.Duration duration1 = Durations.fromSeconds(2);
        com.google.protobuf.Duration duration2 = Durations.fromSeconds(3);
        List<com.google.protobuf.Duration> durations = new ArrayList<>();
        durations.add(duration0);
        durations.add(duration1);
        durations.add(duration2);
        CustomizedBackoff customizedBackoff = CustomizedBackoff.newBuilder().addAllNext(durations).build();
        apache.rocketmq.v2.RetryPolicy retryPolicy = apache.rocketmq.v2.RetryPolicy.newBuilder()
            .setCustomizedBackoff(customizedBackoff).setMaxAttempts(3).build();
        boolean fifo = true;
        int receiveBatchSize = 96;
        com.google.protobuf.Duration longPollingTimeout = Durations.fromSeconds(60);
        Subscription subscription = Subscription.newBuilder().setFifo(fifo).setReceiveBatchSize(receiveBatchSize)
            .setLongPollingTimeout(longPollingTimeout).build();
        Settings settings = Settings.newBuilder().setSubscription(subscription).setBackoffPolicy(retryPolicy).build();
        Resource groupResource = new Resource(FAKE_CONSUMER_GROUP_0);
        ClientId clientId = new ClientId();
        Map<String, FilterExpression> subscriptionExpression = new HashMap<>();
        subscriptionExpression.put(FAKE_TOPIC_0, new FilterExpression("(a > 10 AND a < 100) OR (b IS NOT NULL AND "
            + "b=TRUE)", FilterExpressionType.SQL92));
        final Duration requestTimeout = Duration.ofSeconds(3);
        final PushSubscriptionSettings pushSubscriptionSettings = new PushSubscriptionSettings(clientId,
            fakeEndpoints(), groupResource, requestTimeout, subscriptionExpression);
        pushSubscriptionSettings.sync(settings);
    }
}

File: csharp/tests/PushSubscriptionSettingsTest.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using Castle.Core.Internal;
using Google.Protobuf.WellKnownTypes;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Org.Apache.Rocketmq;
using Proto = Apache.Rocketmq.V2;

namespace tests
{
    [TestClass]
    public class PushSubscriptionSettingsTest
    {
        private const string Namespace = "testNamespace";
        private const string GroupResource = "testConsumerGroup";
        private const string ClientId = "testClientId";
        private const string Endpoint = "127.0.0.1:8080";
        private static readonly TimeSpan RequestTimeout = TimeSpan.FromSeconds(3);
        private static readonly ConcurrentDictionary<string, FilterExpression> SubscriptionExpression =
            new ConcurrentDictionary<string, FilterExpression>(new Dictionary<string, FilterExpression> { { "testTopic", new FilterExpression("*") } });

        private PushSubscriptionSettings CreateSettings()
        {
            return new PushSubscriptionSettings(Namespace, ClientId, new Endpoints(Endpoint), GroupResource, RequestTimeout, SubscriptionExpression);
        }

        [TestMethod]
        public void TestToProtobuf()
        {
            var pushSubscriptionSettings = CreateSettings();
            var settings = pushSubscriptionSettings.ToProtobuf();

            Assert.AreEqual(Proto.ClientType.PushConsumer, settings.ClientType);
            Assert.AreEqual(Duration.FromTimeSpan(RequestTimeout), settings.RequestTimeout);
            Assert.IsFalse(settings.Subscription.Subscriptions.Count == 0);

            var subscription = settings.Subscription;
            Assert.AreEqual(subscription.Group, new Proto.Resource
            {
                ResourceNamespace = Namespace,
                Name = GroupResource
            });

            Assert.IsFalse(subscription.Fifo);

            var subscriptionsList = subscription.Subscriptions;
            Assert.AreEqual(1, subscriptionsList.Count);

            var subscriptionEntry = subscriptionsList[0];
            Assert.AreEqual(Proto.FilterType.Tag, subscriptionEntry.Expression.Type);
            Assert.AreEqual(subscriptionEntry.Topic, new Proto.Resource
            {
                ResourceNamespace = Namespace,
                Name = "testTopic"
            });
        }

        [TestMethod]
        public void TestSync()
        {
            var durations = new List<Duration>
        {
            Duration.FromTimeSpan(TimeSpan.FromSeconds(1)),
            Duration.FromTimeSpan(TimeSpan.FromSeconds(2)),
            Duration.FromTimeSpan(TimeSpan.FromSeconds(3))
        };

            var customizedBackoff = new Proto.CustomizedBackoff
            {
                Next = { durations }
            };

            var retryPolicy = new Proto.RetryPolicy
            {
                CustomizedBackoff = customizedBackoff,
                MaxAttempts = 3
            };

            var subscription = new Proto.Subscription
            {
                Fifo = true,
                ReceiveBatchSize = 96,
                LongPollingTimeout = Duration.FromTimeSpan(TimeSpan.FromSeconds(60))
            };

            var settings = new Proto.Settings
            {
                Subscription = subscription,
                BackoffPolicy = retryPolicy
            };

            var pushSubscriptionSettings = new PushSubscriptionSettings(
                "fakeNamespace", ClientId, new Endpoints(Endpoint), GroupResource, RequestTimeout,
                new ConcurrentDictionary<string, FilterExpression>(SubscriptionExpression));

            pushSubscriptionSettings.Sync(settings);
        }
    }

}